esphome/components/packages/__init__.py

```
@@ -226,32 +238,76 @@ def get_packages(files) -> dict:
+                    packages[idx] = _walk_packages(package_config, callback)
         else:
             raise cv.Invalid(
                 f"Packages must be a key to value mapping or list, got {type(packages)} instead"
             )
+    config[CONF_PACKAGES] = packages
+    return config
+
+
+def do_packages_pass(config: dict, skip_update: bool = False) -> dict:
+    """Processes, downloads and validates all packages in the config.
+    Also extracts and merges all substitutions found in packages into the main config substitutions.
+    """
+    if CONF_PACKAGES not in config:
+        return config
+
+    substitutions = UserDict(config.pop(CONF_SUBSTITUTIONS, {}))
+
+    def process_package_callback(package_config):
+        """This will be called for each package found in the config."""
+        package_config = PACKAGE_SCHEMA(package_config)
+        if isinstance(package_config, str):
+            return package_config  # Jinja string, skip processing
+        if CONF_URL in package_config:
+            package_config = _process_remote_package(package_config, skip_update)
+        # Extract substitutions from the package and merge them into the main substitutions:
+        substitutions.data = merge_config(
+            package_config.pop(CONF_SUBSTITUTIONS, {}), substitutions.data
+        )
+        return package_config
+
+    _walk_packages(config, process_package_callback)
+
+    if substitutions:
+        config[CONF_SUBSTITUTIONS] = substitutions.data
+
+    return config
+
+
+def merge_packages(config: dict) -> dict:
+    """Merges all packages into the main config and removes the `packages:` key."""
+    if CONF_PACKAGES not in config:
+        return config
+
+    # Build flat list of all package configs to merge in priority order:
+    merge_list = []
+
+    def process_package_callback(package_config):
+        """This will be called for each package found in the config."""
+        merge_list.append(package_config)
```

Do we need validation here?  

Hard to know without the typing

---

Added typings and the `Callable` signature to `_walk_packages()`.
You're guaranteed to get a `dict`, which is a package config to be merged.

---

Note to self for later: Need to check what happens if the jinja template evaluates to something like an integer if the error makes sense here

---

You're right, we can run the validation function again (but without allowing Jinja anymore, since it should have been resolved at this stage). I will take a look tonight.

---

Done. So during the merge process we now check that what we are looking to merge at least looks like a package, so if a Jinja expression resolves to something absurd we can fail early.
Added a test to enforce that behavior.